File: LeetCode/110. Balanced Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        class Helper {
            static Res balanced(TreeNode node) {
                if (node == null) return new Res();
                Res l = balanced(node.left);
                Res r = balanced(node.right);
                boolean balanced = l.isBalanced && r.isBalanced && Math.abs(l.depth - r.depth) <= 1;
                return new Res(Math.max(l.depth, r.depth) + 1,
                               balanced);
            }
        }
        return Helper.balanced(root).isBalanced;
    }
} //TC: O(n), SC: O(n)
class Res {
    int depth = 0;
    boolean isBalanced = true;
    Res(int depth, boolean isBalanced) {
        this.depth = depth;
        this.isBalanced = isBalanced;
    }
    Res() {}
}

File: LeetCode/110. Balanced Binary Tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun isBalanced(root: TreeNode?): Boolean {
        return balance(root).isBalanced
    }
    
    fun balance(root: TreeNode?): Res =
        root?.let { //if the node is not null
            val left = balance(root.left) //initate left recursion
            val right = balance(root.right) //initate right recursion
            val isBalanced = left.isBalanced && right.isBalanced && Math.abs(left.depth - right.depth) <= 1 //if both children return they are balance and their depth difference is within 1, the subtree is balanced
            Res(isBalanced, Math.max(left.depth, right.depth)+1)  //return balanced and depth, max from child nodes, plus 1 for proceeding to the next depth
        } ?: Res(true) //return balanced and depth 0 if the node is null
}

class Res(val isBalanced: Boolean, val depth: Int = 0) //for balance function to return both isBalanced and depth